pkg/kv/postgres/store.go

```diff
@@ -0,0 +1,239 @@
+	config, err := pgxpool.ParseConfig(name)
+	if err != nil {
+		return nil, err
+	}
+	pool, err := pgxpool.ConnectConfig(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		// if we return before store uses the pool, free it
+		if pool != nil {
+			pool.Close()
+		}
+	}()
+
+	// acquire connection and make sure we reach the database
+	conn, err := pool.Acquire(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Release()
+	err = conn.Conn().Ping(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	params := parseStoreConfig(config.ConnConfig.RuntimeParams)
+
+	// migrate database
+	err = migrateDatabase(ctx, conn, params)
+	if err != nil {
+		return nil, err
+	}
+	store := &Store{
+		Pool:   pool,
+		Params: params,
+	}
+	pool = nil
+	return store, nil
+}
+
+type Params struct {
+	TableName string
+}
+
+func parseStoreConfig(runtimeParams map[string]string) *Params {
+	p := &Params{
+		TableName: defaultTableName,
+	}
+	if tableName, ok := runtimeParams[paramTableName]; ok {
+		p.TableName = tableName
+	}
+	return p
+}
+
+func migrateDatabase(ctx context.Context, conn *pgxpool.Conn, params *Params) error {
+	_, err := conn.Exec(ctx, `CREATE TABLE IF NOT EXISTS `+params.TableName+` (
+    key BYTEA NOT NULL PRIMARY KEY,
+    value BYTEA NOT NULL);`)
+	return err
+}
+
+func (s *Store) Get(ctx context.Context, key []byte) ([]byte, error) {
+	if key == nil {
+		return nil, kv.ErrMissingKey
+	}
+	row := s.Pool.QueryRow(ctx, `SELECT value FROM `+s.Params.TableName+` WHERE key = $1`, key)
+	var val []byte
+	err := row.Scan(&val)
+	if errors.Is(err, pgx.ErrNoRows) {
+		return nil, kv.ErrNotFound
+	}
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	return val, nil
+}
+
+func (s *Store) Set(ctx context.Context, key, value []byte) error {
+	if key == nil {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+	_, err := s.Pool.Exec(ctx, `INSERT INTO `+s.Params.TableName+`(key,value) VALUES($1,$2)
+			ON CONFLICT (key) DO UPDATE SET value = $2`, key, value)
+	if err != nil {
+		return fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	return nil
+}
+
+func (s *Store) SetIf(ctx context.Context, key, value, valuePredicate []byte) error {
+	if key == nil {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+	var (
+		res pgconn.CommandTag
+		err error
+	)
+	if valuePredicate == nil {
+		// use insert to make sure there was no previous value before
+		res, err = s.Pool.Exec(ctx, `INSERT INTO `+s.Params.TableName+`(key,value) VALUES($1,$2) ON CONFLICT DO NOTHING`, key, value)
+	} else {
+		// update just in case the previous value was same as predicate value
+		res, err = s.Pool.Exec(ctx, `UPDATE `+s.Params.TableName+` SET value=$2 WHERE key=$1 AND value=$3`, key, value, valuePredicate)
+	}
+	if err != nil {
+		return fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	if res.RowsAffected() != 1 {
+		return kv.ErrNotFound
```

Comment:
I don't think that's the proper error. `kv.ErrPredicateFailed` is more appropriate in some cases.

Reply:
I've updated the description to use not found error. I think not found will make the API simple to use - instead of handling two cases: key not found and predicate failed.
WDYT?